bugtool/cmd/helper.go

```diff
@@ -24,6 +24,80 @@ import (
 	"strings"
 )
 
+type tarWriter interface {
+	io.Writer
+	WriteHeader(hdr *tar.Header) error
+}
+
+type walker struct {
+	baseDir, dbgDir string
+	output          tarWriter
+	log             io.Writer
+}
+
+func newWalker(baseDir, dbgDir string, output tarWriter, logger io.Writer) *walker {
+	return &walker{
+		baseDir: baseDir,
+		dbgDir:  dbgDir,
+		output:  output,
+		log:     logger,
+	}
+}
+
+func (w *walker) walkPath(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		fmt.Fprintf(w.log, "Error while walking path %s: %s", path, err)
+		return nil
+	}
+	if info == nil {
+		fmt.Fprintf(w.log, "No file info available")
+		return nil
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		fmt.Fprintf(w.log, "Failed to open %s: %s\n", path, err)
+		// TODO: Write an empty file here, just to hint that this file
+		// existed by there was some problem attempting to add it.
+		return nil
+	}
+	defer file.Close()
+
+	if info.IsDir() {
+		fmt.Fprintf(w.log, "Skipping directory %s\n", info.Name())
+		return nil
+	}
+
+	// Just get the latest fileInfo to make sure that the size is correctly
+	// when the file is write to tar file
+	fpInfo, err := file.Stat()
+	if err != nil {
+		fpInfo, err = os.Lstat(file.Name())
```

Comment:
The only debate I have here is whether we should consider prioritizing copying symlinks as-is. It's likely that with the first attempt to `file.Stat()` a couple of lines up, we will end up copying the full template into the archive multiple times, once for each endpoint; We could consider instead using `Lstat()` directly (or at least first) to first attempt a shallow-copy of symlinks; when later copying the `templates/` directory, we will get the actual template file anyway so there's no big positive to copying the template in multiple times.

On the flip side, if the same binary is copied multiple times into the tar then I'd expect compression of the archive to reduce the size impact of these.

Reply:
I think copying the full template multiple times for now is fine, it's much better than the entire bugtool command failing.